### ISSUE TITLE
Add CXX_STANDARD CMake target property

### ISSUE THREAD

**Issue by feltech** [ada-url/ada#596]
### Version

2.7.4

### Platform

Linux 5.15.0-92-generic #102~20.04.1-Ubuntu SMP Mon Jan 15 13:09:14 UTC 2024 x86_64 x86_64 x86_64 GNU/Linux

### What steps will reproduce the bug?

Using e.g. GCC 9.4, where the default C++ standard is not C++17.

Create a CMake project with a source file that attempts to `#include <ada.h>`.

### How often does it reproduce? Is there a required condition?

Whenever attempting to `#include <ada.h>` without C++17 (or greater) as the target C++ standard version.

### What is the expected behavior?

CMake `target_link_libraries` to `ada::ada` ensures the minimum C++ standard is C++17.

### What do you see instead?

E.g.
```
include/ada/ada_idna.h:71:29: error: ‘string_view’ is not a member of ‘std’
   71 | bool punycode_to_utf32(std::string_view input, std::u32string& out);
      |                             ^~~~~~~~~~~
include/ada/ada_idna.h:71:29: note: ‘std::string_view’ is only available from C++17 onwards
```

### Additional information

On older GCC versions (e.g. 9.4), C++17 is not the default, and so CMake must explicitly add the required compiler flag.

This can be worked around by setting the `CXX_STANDARD` CMake target property in the downstream project. Indeed, Ada's [own tests do this](https://github.com/ada-url/ada/blob/6f0ede327e95d265539bf1ec6f7e130560f25baf/tests/installation/CMakeLists.txt#L5). 

A better fix is to set the appropriate (transitive) target property on Ada itself, either by directly setting `CXX_STANDARD`, or perhaps better, make use of `target_compile_features`.

**Comment by feltech** [comment]
The fix in #597 that closed this issue is not correct. It will not transitively apply the C++ standard requirement to linked CMake targets. 

Apologies if my issue description was not clear on this.

Ada has a test for this that can be updated, as I posted above ([link again](https://github.com/ada-url/ada/blob/6f0ede327e95d265539bf1ec6f7e130560f25baf/tests/installation/CMakeLists.txt#L5)) - though of course it will only flag a problem if the test is compiled with a version of a compiler that does not default to C++17+.

**Comment by lemire** [comment]
@feltech `CXX_STANDARD` applies to all targets created afterward. I am marking this as resolved.

> This property is initialized by the value of the [CMAKE_CXX_STANDARD](https://cmake.org/cmake/help/latest/variable/CMAKE_CXX_STANDARD.html#variable:CMAKE_CXX_STANDARD) variable if it is set when a target is created.

https://cmake.org/cmake/help/latest/prop_tgt/CXX_STANDARD.html

There is no need to mark each individual targets, unless you have different requirements per target.

We could mark C++17 as required. Right now, it will try its best to match C++17. I try to avoid making the standard a requirement as it may artificially break a build. But I admit it is a matter of preferences.

This being said, there is no need to mark each indivial target with compiler features.
